File: frontend/src/api/threatIntel.ts
import { apiClient } from "./client";

export interface ThreatFeedResult {
  id: string;
  source: string;
  reputation_score: number;
  confidence: number;
  threat_classification: string | null;
  raw_data?: any;
  is_cached: boolean;
  created_at: string;
}

export interface Indicator {
  id: string;
  value: string;
  type: string;
  normalized_value: string;
  reputation_score: number;
  confidence_score: number;
  threat_classification: string | null;
  first_seen: string;
  last_seen: string;
  last_updated: string;
  observation_count: number;
  feed_results: ThreatFeedResult[];
}

export const getIndicator = async (value: string, forceRefresh = false): Promise<Indicator> => {
  const { data } = await apiClient.get<Indicator>(`/threat-intel/indicators/${encodeURIComponent(value)}`, {
    params: { force_refresh: forceRefresh }
  });
  return data;
};

export const searchIndicator = async (value: string): Promise<Indicator> => {
  const { data } = await apiClient.post<Indicator>("/threat-intel/indicators/search", { value });
  return data;
};

export const getFeedsStatus = async (): Promise<any> => {
  const { data } = await apiClient.get("/threat-intel/feeds/status");
  return data;
};

File: backend/app/schemas/threat_intel.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, ConfigDict, Field
from uuid import UUID

class ThreatFeedResultBase(BaseModel):
    source: str
    reputation_score: float
    confidence: float
    threat_classification: Optional[str] = None
    is_cached: bool = False
    created_at: datetime
    
class ThreatFeedResultResponse(ThreatFeedResultBase):
    id: UUID
    indicator_id: UUID
    raw_data: Optional[Dict[str, Any]] = None

    model_config = ConfigDict(from_attributes=True)

class IndicatorBase(BaseModel):
    value: str
    type: str
    normalized_value: str

class IndicatorCreate(IndicatorBase):
    pass

class IndicatorResponse(IndicatorBase):
    id: UUID
    reputation_score: float
    confidence_score: float
    threat_classification: Optional[str] = None
    first_seen: datetime
    last_seen: datetime
    last_updated: datetime
    observation_count: int
    
    feed_results: List[ThreatFeedResultResponse] = []

    model_config = ConfigDict(from_attributes=True)

class IndicatorSearchRequest(BaseModel):
    value: str

class IndicatorCorrelationBase(BaseModel):
    source_indicator_id: UUID
    target_indicator_id: UUID
    correlation_type: str
    confidence: float
    created_at: datetime

class IndicatorCorrelationResponse(IndicatorCorrelationBase):
    id: UUID

    model_config = ConfigDict(from_attributes=True)
